csharp/src/Microsoft.ML.OnnxRuntime/ProviderOptions.shared.cs

```diff
@@ -174,13 +192,29 @@ public string GetOptions()
         /// <param name="providerOptions">key/value pairs used to configure a CUDA Execution Provider</param>
         public void UpdateOptions(Dictionary<string, string> providerOptions)
         {
+            var keyStrings = providerOptions.Keys.ToArray();
+            var valStrings = providerOptions.Values.ToArray();
 
-            using (var cleanupList = new DisposableList<IDisposable>())
+            MarshaledStringArray keys = default;
+            MarshaledStringArray values = default;
+            try
             {
-                var keysArray = NativeOnnxValueHelper.ConvertNamesToUtf8(providerOptions.Keys.ToArray(), n => n, cleanupList);
-                var valuesArray = NativeOnnxValueHelper.ConvertNamesToUtf8(providerOptions.Values.ToArray(), n => n, cleanupList);
+                keys = new MarshaledStringArray(keyStrings);
+                values = new MarshaledStringArray(valStrings);
+
+                var nativeKeys = new IntPtr[keyStrings.Length];
+                keys.Fill(nativeKeys);
 
-                NativeApiStatus.VerifySuccess(NativeMethods.OrtUpdateCUDAProviderOptions(handle, keysArray, valuesArray, (UIntPtr)providerOptions.Count));
+                var nativeVals = new IntPtr[valStrings.Length];
+                values.Fill(nativeVals);
+
+                NativeApiStatus.VerifySuccess(NativeMethods.OrtUpdateCUDAProviderOptions(handle, nativeKeys, nativeVals,
+                    (UIntPtr)providerOptions.Count));
+            }
+            finally
+            {
+                keys.Dispose();
+                values.Dispose();
```

Comment:
Can we use a shared implementation for CUDA and TensorRT and SessionOptions.AppendExecutionProvider and pass in the delegate to call in a lambda?